### PR TITLE
Fixed bug regarding unauthorised errors returning 302 instead of 401 when HtmlRedirect is null

### PR DESCRIPTION
**Pull request by bumblebeeman**
As discussed [here](https://github.com/ServiceStack/Issues/issues/363)

### CODE REVIEW

File: src/ServiceStack.Mvc/ServiceStackController.cs [hunk 1]
```diff
@@ -229,10 +250,6 @@ protected virtual TUserSession SessionAs<TUserSession>()
         {
             return ServiceStackProvider.SessionAs<TUserSession>();
         }
-        protected virtual void SaveSession(IAuthSession session, TimeSpan? expiresIn = null)
-        {
-            ServiceStackProvider.Request.SaveSession(session, expiresIn);
-        }
```
Comment (bumblebeeman) [inline src/ServiceStack.Mvc/ServiceStackController.cs]: Was working of code from a couple of days ago, I'm guessing you added this after

File: src/ServiceStack.Mvc/ServiceStackController.cs [hunk 2]
```diff
@@ -53,7 +53,17 @@ public virtual ActionResult AuthenticationErrorResult
...
+
+                var encoded = returnUrl.UrlEncode();
+                var unauthorizedRedirectUrl = UnauthorizedRedirectUrl;
+
+                if (string.IsNullOrEmpty(unauthorizedRedirectUrl))
+                {
+                    throw new HttpException(401, "Unauthorized");
+                }
+
+                var formatted = unauthorizedRedirectUrl.Fmt(encoded);
```
Comment (Nness) [inline src/ServiceStack.Mvc/ServiceStackController.cs]: You removed `?redirect={0}#f=Unauthorized`, where you using string.format encoded after.. Even though no error will be throwing, but what's the point? And by this change you removed default behavior. 

Things can be improve is under AuthFeature class's GetHtmlRedirect.
```
if (feature != null)
                return feature.HtmlRedirect;
```
to 
```
if (feature != null && feature.HtmlRedirect != null)
                return feature.HtmlRedirect;
```
Also HtmlRedirect did setup automatically, unless you change that value after AuthFeature been created.

Comment (bumblebeeman) [inline src/ServiceStack.Mvc/ServiceStackController.cs]: Ok, I did raise this as an issue first, as I wasn't sure.

The main point is, in the documentation it says that if you set HtmlRedirect to null, you will get a 401 response when unauthorised, instead you get a 302, which I don't want.

I only want a 302 when I set HtmlRedirect to an actual value, with the temporary redirection being to that value.

**Comment by Nness** [comment]
Well I'm not author, just curious about this pull request.

I misunderstand part of your fix I think. My proposal is wrong.

One way of fix without changing behavior might be following. 

```
        public virtual ActionResult AuthenticationErrorResult
        {
            get
            {
                var returnUrl = HttpContext.Request.GetPathAndQuery();
                var unauthorizedRedirectUrl = UnauthorizedRedirectUrl;
                if (unauthorizedRedirectUrl.IsNullOrEmpty() || unauthorizedRedirectUrl.StartsWith("?redirect")) {
                    return new HttpUnauthorizedResult("Unauthorized");
                }
                return new RedirectResult(unauthorizedRedirectUrl.Fmt(returnUrl.UrlEncode()));
            }
        }
```

What do you think? This way it won't change original behavior. of `UnauthorizedRedirectUrl`.

**Comment by bumblebeeman** [comment]
Ok, I guess I will wait and see if this gets accepted as a pull request, and if not submit another one, it's a fairly minor change so might be easier for the devs (mythz?) to pull and then patch.

One thought, I prefer throwing HttpExceptions, as there is no *HttpForbiddenResult*, so this keeps it consistent with the code below.

**Comment by mythz** [comment]
Hey everyone I've just added a commit to clean up the PR with both your suggestions, please let me know if there are other changes you would like to make.

**Pull request merged**